### PR TITLE
gamma distributions

### PR DESCRIPTION
**Pull request by Jules-Deschamps**

### CODE REVIEW

**Review by alebrigant — CHANGES_REQUESTED** [review]
Great work ! There is a lot of code to review, therefore a lot of comments. Mostly style though !

File: geomstats/information_geometry/gamma.py [hunk 1]
```diff
@@ -0,0 +1,1131 @@
...
+
+            Returns
+            -------
+            pdf_at_x : array-like, shape=[..., n_points]
+                Values of pdf at x for each value of the parameters provided
+                by point.
+            """
+            pdf_at_x = gs.array(
+                [
+                    gamma.pdf(t, a=point[:, 0], scale=1 / point[:, 1])
```
[the remark this answers is not in the document]
Comment (Jules-Deschamps) [reply geomstats/information_geometry/gamma.py:178]: You're right I was still in natural coordinates but it makes more sense if everything is in standard coordinates.

File: geomstats/information_geometry/gamma.py [hunk 2]
```diff
@@ -0,0 +1,1198 @@
...
+                    for t in gs.array(x)
+                ]
+            )
+            pdf_at_x = gs.squeeze(gs.stack(pdf_at_x, axis=0)).transpose()
+
+            return pdf_at_x
+
+        return pdf
+
+    def natural_to_standard_point(self, point):
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:195]: Remove '_point' in name to be consistent with Hypersphere.

File: geomstats/information_geometry/gamma.py [hunk 3]
```diff
@@ -0,0 +1,1198 @@
...
+        point : array-like, shape=[..., 2]
+            Point of the Gamma manifold, given in standard coordinates.
+        """
+        point = gs.to_ndarray(point, to_ndim=2)
+
+        point[:, 1] = point[:, 0] / point[:, 1]
+
+        return gs.squeeze(point)
+
+    def standard_to_natural_point(self, point):
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:216]: Ditto.

File: geomstats/information_geometry/gamma.py [hunk 4]
```diff
@@ -0,0 +1,1198 @@
...
+            Point of the Gamma manifold, given in standard coordinates.
+
+        Returns
+        -------
+        point : array-like, shape=[..., 2]
+            Point of the Gamma manifold, given in natural coordinates.
+        """
+        return self.natural_to_standard_point(point)
+
+    def natural_to_standard_vec(self, vec, base_point):
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:233]: Rename to `tangent_natural_to_standard` to be consistent with Hypersphere.

File: geomstats/information_geometry/gamma.py [hunk 5]
```diff
@@ -0,0 +1,1198 @@
...
+            [[gs.ones(n_points), gs.zeros(n_points)], [1 / scale, -kappa / scale**2]]
+        )
+        jac = gs.transpose(jac, [2, 0, 1])
+
+        vec = vec.reshape((n_points, 2, 1))
+        vec = gs.matmul(jac, vec)
+
+        return gs.squeeze(vec)
+
+    def standard_to_natural_vec(self, point):
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:269]: Ditto.

File: geomstats/information_geometry/gamma.py [hunk 6]
```diff
@@ -0,0 +1,1198 @@
...
+        """
+        return self.natural_to_standard_vec(point)
+
+
+class GammaMetric(RiemannianMetric):
+    """Class for the Fisher information metric on Gamma distributions.
+
+    References
+    ----------
+    Arwini, K. A., & Dodson, C. T. (2008). Information geometry (pp. 31-54).
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:294]: This is still not in the same format as e.g. [here](https://github.com/geomstats/geomstats/blob/9bad7b8cc618fc0b9cc3dee13d01698a8f86bb11/geomstats/geometry/spd_matrices.py#L483). It must start with `.. [AD2008]` etc.

File: geomstats/information_geometry/gamma.py [hunk 7]
```diff
@@ -0,0 +1,1131 @@
...
+
+    def christoffels(self, base_point):
+        """Compute the Christoffel symbols, when given in simple coordinates.
+
+        Compute the Christoffel symbols of the Fisher information metric.
+
+        References
+        ----------
+        Arwini, K. A., & Dodson, C. T. (2008). Information geometry (pp. 31-54).
+        Springer Berlin Heidelberg.
```
[the remark this answers is not in the document]
Comment (alebrigant) [reply geomstats/information_geometry/gamma.py:342]: Same as previous comment.

File: geomstats/information_geometry/gamma.py [hunk 8]
```diff
@@ -0,0 +1,1198 @@
...
+        For computation purposes, we replace the value of
+        (gs.polygamma(1, x) - 1/x) by an equivalent (close lower-bound) when it becomes
+        too difficult to compute, as per in the second reference.
+
+        References
+        ----------
+        Arwini, K. A., & Dodson, C. T. (2008). Information geometry (pp. 31-54).
+        Springer Berlin Heidelberg.
+
+        Guo, B. N., Qi, F., Zhao, J. L., & Luo, Q. M. (2015). Sharp inequalities for
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:344]: Ditto.

File: geomstats/information_geometry/gamma.py [hunk 9]
```diff
@@ -0,0 +1,1198 @@
...
+        Compute the Jacobian of the Christoffel symbols of the
+        Fisher information metric.
+
+        For computation purposes, we replace the value of
+        (gs.polygamma(1, x) - 1/x) and (gs.polygamma(2,x) + 1/x**2) by an equivalent
+        (close bounds) when they become too difficult to compute.
+
+        References
+        ----------
+        Guo, B. N., Qi, F., Zhao, J. L., & Luo, Q. M. (2015). Sharp inequalities for
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:429]: Same as previous comments about reference format.

File: geomstats/information_geometry/gamma.py [hunk 10]
```diff
@@ -0,0 +1,1198 @@
...
+            Tangent vector at base point.
+        base_point : array-like, shape=[..., dim]
+            Base point.
+        n_steps : int
+            Number of steps for integration.
+            Optional, default: 100.
+        step : str, {'euler', 'rk4'}
+            Numerical scheme to use for integration.
+            Optional, default: 'euler'.
+        method : string, {'geomstats', 'lsoda'}
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:535]: Here I think `solver` would make more sense than `method`.

File: geomstats/information_geometry/gamma.py [hunk 11]
```diff
@@ -0,0 +1,1198 @@
...
+        self,
+        point,
+        base_point,
+        n_steps=N_STEPS,
+        init=None,
+        step="euler",
+        max_iter=25,
+        verbose=False,
+        tol=gs.atol,
+        solver="geodesic_shooting",
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:562]: I think there was a mix up: here `method` makes more sense than `solver`, which refers to solving ODEs. See previous comment.

**Review by alebrigant — CHANGES_REQUESTED** [review]
Some more comments left to address !

File: tests/data/gamma_data.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+            dict(point=point, tangent_vecs=tangent_vecs, exp_method="ivp"),
+        ]
+        return self.generate_tests([], random_data)
+
+    def exp_control_test_data(self):
+        n_times = 10
+        base_point = self.space().random_point(n_times)
+        tangent_vec = self.space().metric.random_unit_tangent_vec(
+            base_point=base_point, n_vectors=1
+        )
```
[the remark this answers is not in the document]
Comment (alebrigant) [reply tests/data/gamma_data.py:274]: Can we test both solvers with tangent vectors of different norms, not only lsoda ? And maybe select a norm randomly in a set of integers (e.g. `random.sample(range(1, 5))`) ?

File: tests/data/gamma_data.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+            ),
+        ]
+        return self.generate_tests([], random_data)
+
+    def exp_after_log_control_test_data(self):
+        n_times = 10
+        base_point = self.space().random_point(n_times)
+        tangent_vec = self.space().metric.random_unit_tangent_vec(
+            base_point=base_point, n_vectors=1
+        )
```
[the remark this answers is not in the document]
Comment (alebrigant) [reply tests/data/gamma_data.py:312]: Same.

File: tests/tests_geomstats/test_gamma.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_autograd_only
+    def test_exp_control(self, base_point, tangent_vec, exp_solver):
+        """Test exp at a random base point for a tangent vector of controlled norm."""
+        end_point = self.metric().exp(
+            tangent_vec=tangent_vec, base_point=base_point, solver=exp_solver
+        )
+        result = gs.any(gs.isnan(end_point))
+        self.assertAllClose(result, False)
```
Comment (alebrigant) [inline tests/tests_geomstats/test_gamma.py:127]: Can we also test exp and log on points where the geodesic is known, or at least its shape ? e.g. if vertical lines are geodesics, then `exp` of a base point and a vertical vector should yield a point vertically aligned with the base point, and `log` of two vertically aligned points should yield a vertial vector.

File: geomstats/information_geometry/gamma.py [hunk 15]
```diff
@@ -0,0 +1,1200 @@
...
+            initial conditions.
+        """
+        if solver == "geomstats":
+            return super(GammaMetric, self).geodesic(
+                initial_point, end_point, initial_tangent_vec
+            )
+        if solver == "vp":
+            return self.geodesic_vp(initial_point, end_point, initial_tangent_vec)
+
+    def _geodesic_ivp(self, initial_point, initial_tangent_vec, n_steps=N_STEPS):
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:661]: This function should be tested, see e.g. [here](https://github.com/geomstats/geomstats/blob/9bad7b8cc618fc0b9cc3dee13d01698a8f86bb11/tests/tests_geomstats/test_dirichlet.py#L159).

File: geomstats/information_geometry/gamma.py [hunk 16]
```diff
@@ -0,0 +1,1200 @@
...
+                    initial_state = gs.hstack([point, vec])
+                    solution = odeint(ivp, initial_state, t_int, ())
+                    geod.append(solution[:, : self.dim])
+
+            geod = geod[0] if len(initial_point) == 1 else gs.stack(geod)
+            return gs.where(geod < gs.atol, gs.atol, geod)
+
+        return path
+
+    def _exp_ivp(self, tangent_vec, base_point, n_steps=N_STEPS):
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:749]: This function should be tested.

File: geomstats/information_geometry/gamma.py [hunk 17]
```diff
@@ -0,0 +1,1200 @@
...
+            End point of the geodesic starting at base_point with
+            initial velocity tangent_vec and stopping at time 1.
+        """
+        stop_time = 1.0
+        geodesic = self._geodesic_ivp(base_point, tangent_vec, n_steps)
+        exp = geodesic(stop_time)
+
+        return exp
+
+    def _approx_geodesic_bvp(
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:778]: This function should be tested, see e.g. [here](https://github.com/geomstats/geomstats/blob/9bad7b8cc618fc0b9cc3dee13d01698a8f86bb11/tests/tests_geomstats/test_dirichlet.py#L216).

File: geomstats/information_geometry/gamma.py [hunk 18]
```diff
@@ -0,0 +1,1200 @@
...
+        dim = initial_point.shape[0]
+        x0 = gs.ones(dim * (degree - 1))
+        jac = jacobian if jac_on else None
+        sol = minimize(f2minimize, x0, method=method, jac=jac)
+        opt_param = sol.x.reshape((dim, degree - 1)).T
+        _, dist, curve, velocity = cost_fun(opt_param)
+
+        return dist, curve, velocity
+
+    def _geodesic_bvp(
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:928]: This function should be tested, see e.g. [here](https://github.com/geomstats/geomstats/blob/9bad7b8cc618fc0b9cc3dee13d01698a8f86bb11/tests/tests_geomstats/test_dirichlet.py#L167).

File: geomstats/information_geometry/gamma.py [hunk 19]
```diff
@@ -0,0 +1,1200 @@
...
+            initial_point=base_point,
+            end_point=point,
+            jacobian=jacobian,
+        )
+        geodesic_at_t = geodesic(t)
+        log = n_steps * (geodesic_at_t[..., 1, :] - geodesic_at_t[..., 0, :])
+
+        return gs.squeeze(gs.stack(log))
+
+    def geodesic_vp(
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:1145]: Ditto.

**Review by alebrigant — CHANGES_REQUESTED** [review]
Thanks for the modifications ! We're almost there: just a few more tests to add so that the coverage is good.

File: tests/data/gamma_data.py [hunk 20]
```diff
@@ -0,0 +1,619 @@
...
+            ),
+        ]
+        return self.generate_tests([], random_data)
+
+    def geodesic_shape_test_data(self):
+        random_data = [
+            dict(
+                point=self.space().random_point(),
+                n_vec=1,
+                norm=random.choice(self.n_points_list),
```
Comment (alebrigant) [inline tests/data/gamma_data.py:544]: Same.

File: tests/data/gamma_data.py [hunk 21]
```diff
@@ -0,0 +1,619 @@
...
+                expected=(4, 10, 2),
+            ),
+        ]
+        return self.generate_tests([], random_data)
+
+    def gamma_level_geodesics_are_horizontal_test_data(self):
+        random_data = [
+            dict(
+                base_point=self.space().random_point(),
+                norm=random.choice(self.n_points_list),
```
Comment (alebrigant) [inline tests/data/gamma_data.py:596]: Same.

File: geomstats/information_geometry/gamma.py [hunk 22]
```diff
@@ -0,0 +1,1190 @@
...
```
Comment (alebrigant) [inline geomstats/information_geometry/gamma.py:1135]: Rename to `_geodesic_vp`.

**Review by alebrigant — CHANGES_REQUESTED** [review]
Last changes:
- add `test_log_after_exp`
- remove duplicates in random tests
- add a lower bound in `random_point` and use it to improve performance of test of horizontal geodesics
- other comments to be addressed are in the code.

**Pull request merged**